File: backend/app/models/wallet.py
"""
钱包和交易相关数据模型
"""
from datetime import datetime
from app import db


class UserWallet(db.Model):
    """用户钱包模型"""
    __tablename__ = 'user_wallets'
    
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False, unique=True)
    wallet_address = db.Column(db.String(42), nullable=False, unique=True)
    private_key_encrypted = db.Column(db.Text, nullable=False)
    crawler_coin_balance = db.Column(db.BigInteger, default=1000000)  # wei单位
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # 关系 - 通过user_id关联
    user = db.relationship('User', backref='wallet')
    
    def __repr__(self):
        return f'<UserWallet {self.wallet_address}>'
    
    @property
    def balance_in_tokens(self):
        """以代币单位显示余额"""
        return self.crawler_coin_balance / 1000000  # 假设6位小数
    
    def to_dict(self):
        """转换为字典"""
        return {
            'id': self.id,
            'user_id': self.user_id,
            'wallet_address': self.wallet_address,
            'crawler_coin_balance': self.crawler_coin_balance,
            'balance_in_tokens': self.balance_in_tokens,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }


class TokenTransaction(db.Model):
    """代币交易记录模型"""
    __tablename__ = 'token_transactions'
    
    id = db.Column(db.Integer, primary_key=True)
    from_user_id = db.Column(db.Integer, db.ForeignKey('users.id'), index=True)
    to_user_id = db.Column(db.Integer, db.ForeignKey('users.id'), index=True)
    transaction_type = db.Column(db.String(50), nullable=False, index=True)
    amount = db.Column(db.BigInteger, nullable=False)
    related_id = db.Column(db.Integer)
    related_type = db.Column(db.String(50))
    blockchain_tx_hash = db.Column(db.String(66))
    status = db.Column(db.String(20), default='pending', index=True)
    description = db.Column(db.Text)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    def __repr__(self):
        return f'<TokenTransaction {self.transaction_type} {self.amount}>'
    
    @property
    def amount_in_tokens(self):
        """以代币单位显示金额"""
        return self.amount / 1000000  # 假设6位小数
    
    def to_dict(self):
        """转换为字典"""
        return {
            'id': self.id,
            'from_user_id': self.from_user_id,
            'to_user_id': self.to_user_id,
            'transaction_type': self.transaction_type,
            'amount': self.amount,
            'amount_in_tokens': self.amount_in_tokens,
            'related_id': self.related_id,
            'related_type': self.related_type,
            'blockchain_tx_hash': self.blockchain_tx_hash,
            'status': self.status,
            'description': self.description,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }


class TipRecord(db.Model):
    """打赏记录模型"""
    __tablename__ = 'tip_records'
    
    id = db.Column(db.Integer, primary_key=True)
    from_user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False, index=True)
    to_user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False, index=True)
    target_type = db.Column(db.String(20), nullable=False, index=True)
    target_id = db.Column(db.Integer, nullable=False)
    amount = db.Column(db.BigInteger, nullable=False)
    transaction_id = db.Column(db.Integer, db.ForeignKey('token_transactions.id'))
    blockchain_tx_hash = db.Column(db.String(66))
    message = db.Column(db.Text)
    is_anonymous = db.Column(db.Boolean, default=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    
    # 关系
    transaction = db.relationship('TokenTransaction', backref='tip_record')
    
    def __repr__(self):
        return f'<TipRecord {self.target_type}:{self.target_id} {self.amount}>'
    
    @property
    def amount_in_tokens(self):
        """以代币单位显示金额"""
        return self.amount / 1000000  # 假设6位小数
    
    def to_dict(self):
        """转换为字典"""
        return {
            'id': self.id,
            'from_user_id': self.from_user_id,
            'to_user_id': self.to_user_id,
            'target_type': self.target_type,
            'target_id': self.target_id,
            'amount': self.amount,
            'amount_in_tokens': self.amount_in_tokens,
            'transaction_id': self.transaction_id,
            'blockchain_tx_hash': self.blockchain_tx_hash,
            'message': self.message,
            'is_anonymous': self.is_anonymous,
            'created_at': self.created_at.isoformat() if self.created_at else None
        }

File: frontend/types/wallet.ts
/**
 * Web3钱包系统相关的类型定义
 */

// 用户钱包
export interface UserWallet {
  id: number;
  user_id: number;
  wallet_address: string;
  private_key_encrypted: string;
  crawler_coin_balance: string; // 使用字符串存储大数值
  created_at: string;
  updated_at: string;
}

// 钱包创建请求
export interface CreateWalletRequest {
  user_id: number;
  password?: string; // 可选的额外密码保护
}

// 钱包信息响应
export interface WalletInfo {
  wallet_address: string;
  crawler_coin_balance: string;
  balance_formatted: string; // 格式化后的余额显示
  created_at: string;
}

// 代币交易类型
export enum TransactionType {
  TIP = 'tip',
  PURCHASE_STORAGE = 'purchase_storage',
  POST_COST = 'post_cost',
  REPLY_COST = 'reply_cost',
  INITIAL_GRANT = 'initial_grant',
  TRANSFER = 'transfer',
}

// 代币交易记录
export interface TokenTransaction {
  id: number;
  from_user_id?: number;
  to_user_id?: number;
  transaction_type: TransactionType;
  amount: string; // 使用字符串存储大数值
  related_id?: number;
  related_type?: string;
  blockchain_tx_hash?: string;
  status: 'pending' | 'confirmed' | 'failed';
  description?: string;
  created_at: string;
  updated_at: string;
  
  // 关联数据
  from_user?: {
    id: number;
    username: string;
    avatar_url: string;
  };
  to_user?: {
    id: number;
    username: string;
    avatar_url: string;
  };
}

// 创建交易请求
export interface CreateTransactionRequest {
  from_user_id?: number;
  to_user_id?: number;
  transaction_type: TransactionType;
  amount: string;
  related_id?: number;
  related_type?: string;
  description?: string;
}

// 打赏记录
export interface TipRecord {
  id: number;
  from_user_id: number;
  to_user_id: number;
  target_type: 'post' | 'reply' | 'comment';
  target_id: number;
  amount: string;
  transaction_id?: number;
  blockchain_tx_hash?: string;
  message?: string;
  is_anonymous: boolean;
  created_at: string;
  
  // 关联数据
  from_user?: {
    id: number;
    username: string;
    avatar_url: string;
  };
  to_user?: {
    id: number;
    username: string;
    avatar_url: string;
  };
  transaction?: TokenTransaction;
}

// 创建打赏请求
export interface CreateTipRequest {
  to_user_id: number;
  target_type: 'post' | 'reply' | 'comment';
  target_id: number;
  amount: string;
  message?: string;
  is_anonymous?: boolean;
}

// 钱包操作响应
export interface WalletOperationResponse {
  success: boolean;
  message: string;
  data?: any;
  transaction_id?: number;
  blockchain_tx_hash?: string;
}

// 代币余额信息
export interface TokenBalance {
  balance: string; // wei单位
  balance_formatted: string; // 格式化显示
  balance_ether: string; // ether单位
}

// 钱包统计信息
export interface WalletStats {
  total_earned: string;
  total_spent: string;
  total_tips_received: string;
  total_tips_sent: string;
  transaction_count: number;
}

// 钱包配置
export interface WalletConfig {
  initial_balance: string; // 初始余额（wei）
  post_cost: string; // 发帖费用（wei）
  reply_cost: string; // 回复费用（wei）
  storage_price_per_mb: string; // 每MB存储价格（wei）
  min_tip_amount: string; // 最小打赏金额（wei）
  max_tip_amount: string; // 最大打赏金额（wei）
}

// 常量定义
export const WALLET_CONSTANTS = {
  INITIAL_BALANCE: '100000000000000000000', // 100 CRAWLER Coin (wei) - 调整为PostgreSQL bigint安全范围
  POST_COST: '1000000000000000000', // 1 CRAWLER Coin (wei)
  REPLY_COST: '500000000000000000', // 0.5 CRAWLER Coin (wei)
  STORAGE_PRICE_PER_MB: '10000000000000000000', // 10 CRAWLER Coin per MB (wei)
  MIN_TIP_AMOUNT: '100000000000000000', // 0.1 CRAWLER Coin (wei)
  MAX_TIP_AMOUNT: '100000000000000000000', // 100 CRAWLER Coin (wei) - 调整为安全范围
  DECIMALS: 18,
  SYMBOL: 'CRAWLER',
  NAME: 'CRAWLER Coin',
} as const;

// 工具函数类型
export interface WalletUtils {
  formatBalance: (balance: string) => string;
  parseBalance: (balance: string) => string;
  validateAmount: (amount: string) => boolean;
  generateWallet: () => Promise<{ address: string; privateKey: string }>;
  encryptPrivateKey: (privateKey: string, password: string) => string;
  decryptPrivateKey: (encryptedKey: string, password: string) => string;
}
